dspinner/pin.go

```diff
@@ -0,0 +1,1069 @@
+// stores pin data in a datastore.
+package dspinner
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	ds "github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/query"
+	ipfspinner "github.com/ipfs/go-ipfs-pinner"
+	"github.com/ipfs/go-ipfs-pinner/dsindex"
+	"github.com/ipfs/go-ipfs-pinner/ipldpinner"
+	ipld "github.com/ipfs/go-ipld-format"
+	logging "github.com/ipfs/go-log"
+	mdag "github.com/ipfs/go-merkledag"
+	"github.com/polydawn/refmt/cbor"
+)
+
+const (
+	loadTimeout    = 5 * time.Second
+	rebuildTimeout = 2 * time.Minute
+
+	pinKeyPath   = "/.pins/pin"
+	indexKeyPath = "/.pins/index"
+	dirtyKeyPath = "/.pins/state/dirty"
+)
+
+var (
+	// ErrNotPinned is returned when trying to unpin items that are not pinned.
+	ErrNotPinned = fmt.Errorf("not pinned or pinned indirectly")
+
+	log logging.StandardLogger = logging.Logger("pin")
+
+	linkDirect, linkRecursive string
+
+	pinCidDIndexPath string
+	pinCidRIndexPath string
+	pinNameIndexPath string
+
+	pinDatastoreKey = ds.NewKey("/local/pins")
+
+	dirtyKey = ds.NewKey(dirtyKeyPath)
+)
+
+func init() {
+	directStr, ok := ipfspinner.ModeToString(ipfspinner.Direct)
+	if !ok {
+		panic("could not find Direct pin enum")
+	}
+	linkDirect = directStr
+
+	recursiveStr, ok := ipfspinner.ModeToString(ipfspinner.Recursive)
+	if !ok {
+		panic("could not find Recursive pin enum")
+	}
+	linkRecursive = recursiveStr
+
+	pinCidRIndexPath = path.Join(indexKeyPath, "cidRindex")
+	pinCidDIndexPath = path.Join(indexKeyPath, "cidDindex")
+	pinNameIndexPath = path.Join(indexKeyPath, "nameIndex")
+}
+
+// pinner implements the Pinner interface
+type pinner struct {
+	lock sync.RWMutex
+
+	dserv  ipld.DAGService
+	dstore ds.Datastore
+
+	cidDIndex dsindex.Indexer
+	cidRIndex dsindex.Indexer
+	nameIndex dsindex.Indexer
+
+	dirty uint64
+}
+
+var _ ipfspinner.Pinner = (*pinner)(nil)
+
+type pin struct {
+	id       string
+	cid      cid.Cid
+	metadata map[string]interface{}
+	mode     ipfspinner.Mode
+	name     string
+}
+
+func (p *pin) dsKey() ds.Key {
+	return ds.NewKey(path.Join(pinKeyPath, p.id))
+}
+
+func newPin(c cid.Cid, mode ipfspinner.Mode, name string) *pin {
+	return &pin{
+		id:   ds.RandomKey().String(),
+		cid:  c,
+		name: name,
+		mode: mode,
+	}
+}
+
+type syncDAGService interface {
+	ipld.DAGService
+	Sync() error
+}
+
+// New creates a new pinner using the given datastore as a backend
+func New(dstore ds.Datastore, serv ipld.DAGService) ipfspinner.Pinner {
+	return &pinner{
+		cidDIndex: dsindex.New(dstore, pinCidDIndexPath),
+		cidRIndex: dsindex.New(dstore, pinCidRIndexPath),
+		nameIndex: dsindex.New(dstore, pinNameIndexPath),
+		dserv:     serv,
+		dstore:    dstore,
+	}
+}
+
+// Pin the given node, optionally recursive
+func (p *pinner) Pin(ctx context.Context, node ipld.Node, recurse bool) error {
+	err := p.dserv.Add(ctx, node)
+	if err != nil {
+		return err
+	}
+
+	c := node.Cid()
+	cidStr := c.String()
```

Comment:
I agree, but since the CID is used as part of the path that makes up an index, this cannot have arbitrary bytes (due to need to have path separator: `cidindex/cid_string/pin_id`

Probably the most correct solution is for the indexer to base64-URL encode the index and key values.  Then arbitrary strings can be used for both.

Will defer to follow-on PR.

Reply:
Done.

Base36 encode the index and strings to allow them to contain any characters  without interfering with the datastore key path.  Base36 was chosen because it is slightly more compact than Base32, more portable than Base58, Base64, etc., and because it has a very fast implementation.

dspinner can now use cid.KeyString() to store the raw byte string as in index. This avoids having to encode the cid every time it is used as an index.

Reply:
After talking this over a bit with @gammazero @Stebalien and @ribasushi it seems like we're probably better off and still safe using Base64URL encoding here. It's less bytes on disk and AFAICT every datastore we work with except for FlatFS supports case-sensitive data, and FlatFS is only used for data blocks.

If it turns out we regret this decision we can always migrate to something else, but for now it seems like doing the fast and efficient thing seems like the best approach.

Reply:
Done
Base64 it is!!